Mathlib/RingTheory/NoetherNormalization.lean

```diff
@@ -0,0 +1,253 @@
+/-
+Copyright (c) 2025 Sihan Su. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca, Sihan Su, Wan Lin, Xiaoyang Su
+-/
+import Mathlib.Algebra.MvPolynomial.Monad
+import Mathlib.Data.List.Indexes
+import Mathlib.RingTheory.IntegralClosure.IsIntegralClosure.Basic
+/-!
+# Noether Normalizaton
+This file contains a proof of Noether normalization lemma by Nagata.
+
+## Main Results
+Let `A` be a finitely generated algebra over a field `k`.
+Then there exists natural number `r` and an injective homomorphism
+from `k[X_1, X_2, ..., X_r]` to `A` such that `A` is integral over `k[X_1, X_2, ..., X_r]`.
+
+## Reference
+* <https://stacks.math.columbia.edu/tag/00OW>
+
+-/
+open Polynomial MvPolynomial Ideal BigOperators Nat RingHom List
+
+variable {k : Type*} [Field k] {n : ℕ} (f : MvPolynomial (Fin (n + 1)) k)
+variable (v w : Fin (n + 1) →₀ ℕ)
+
+namespace NoetherNormalization
+
+/-- Suppose `f` is a nonzero polynomial of `n+1` variables : `X_0,...,X_n`.
+`up` is a number which is big enough. -/
+local notation3 "up" => 2 + f.totalDegree
+
+/-- `r` maps i to `up ^ i`-/
+local notation3 "r" => fun (i : Fin (n + 1)) ↦ up ^ i.1
+
+/-- We construct an algebra map `T1` which maps `X_i` into `X_i + X_0 ^ r_i` when `i ≠ 0`,
+and maps `f` into some monic polynomial (regarded as a polynomial of `X_0`).-/
+noncomputable abbrev T1 (c : k) :
+    MvPolynomial (Fin (n + 1)) k →ₐ[k] MvPolynomial (Fin (n + 1)) k :=
+  aeval fun i ↦ if i = 0 then X 0 else X i + c • X 0 ^ r i
+
+private lemma inv_pair (c : k) :
+    (T1 f c).comp (T1 f (-c)) = AlgHom.id _ _ := by
+  rw [comp_aeval, ← MvPolynomial.aeval_X_left]
+  ext i v
+  cases i using Fin.cases <;>
+  simp [Fin.succ_ne_zero]
+
+/-- `T1` leads to an algebra equiv `T`.-/
+private noncomputable abbrev T := AlgEquiv.ofAlgHom (T1 f 1) (T1 f (-1))
+  (inv_pair f 1) (by convert (inv_pair f (-1)); exact (InvolutiveNeg.neg_neg 1).symm)
```

Comment:
```suggestion
  (inv_pair f 1) ((by simpa using inv_pair f (-1)))
```

Reply:
The outer pair of parens around `by simpa ...` is surely unnecessary.